feat(vrl): add a max_depth parameter to the parse_json function

nabokihms opened the pull request:
Add the max_depth option to the `parse_json` VRL function to partially traverse highly nested JSON-formatted messages.

These changes were tested in the Kubernetes cluster 1.21 on top of the Vector 0.20

```
$ parse_json!("{\"1\": {\"2\": {\"3\": {\"4\": {\"5\": {\"6\": \"finish\"}}}}}}", max_depth: 5)

{ "1": { "2": { "3": { "4": { "5": "{\"6\": \"finish\"}" } } } } }
```
```
$ parse_json!("{\"1\": {\"2\": {\"3\": {\"4\": {\"5\": {\"6\": \"finish\"}}}}}}", max_depth: 4)

{ "1": { "2": { "3": { "4": "{\"5\": {\"6\": \"finish\"}}" } } } }
```
```
$ parse_json!("{\"1\": {\"2\": {\"3\": {\"4\": {\"5\": {\"6\": \"finish\"}}}}}}", max_depth: 10)

{ "1": { "2": { "3": { "4": {"5": {"6": "finish"} } } } } }
```
```
$ parse_json!("{\"1\": {\"2\": {\"3\": {\"4\": {\"5\": {\"6\": \"finish\"}}}}}}", max_depth: 129)

function call error for "parse_json" at (0:95): max_depth value should be greater than 0 and less than 128, got 129
```
***
Closes #12458
Signed-off-by: m.nabokikh <maksim.nabokikh@flant.com>

JeanMertz reviewed (commented):
Thank you! This looks great. I left one comment on how we could help performance for the most-common (non-max-depth) case. We also need to update the `parse_json.cue` docs to document this new feature.

nabokihms commented:
@JeanMertz hope I get your idea right. 

I would also like to add benchmarks for the max_depth function. Is the `benches` folder the right place?

JeanMertz commented:
> @JeanMertz hope I got your idea right.

Looks great!

> I would also like to add benchmarks for the max_depth function. Is the `benches` folder the right place?

Correct! We have a macro `bench_function` that is used in `lib/vrl/stdlib/benches/benches.rs`, which should help you get going. I assume we already have one for `parse_json` in there.

nabokihms commented:
That's the result of the benchmark.
```
vrl_stdlib/functions/parse_json/map                                                                            
                        time:   [451.52 ns 462.79 ns 474.89 ns]
                        thrpt:  [2.1058 Melem/s 2.1608 Melem/s 2.2148 Melem/s]

vrl_stdlib/functions/parse_json/map_max_depth                                                                             
                        time:   [1.1024 us 1.1209 us 1.1374 us]
                        thrpt:  [879.22 Kelem/s 892.18 Kelem/s 907.11 Kelem/s]

```
I assume that this PR is ready for the second round of review 😃

JeanMertz commented:
It is unfortunate that the `max_depth` version is about 2,5x slower. I don't see any direct ways to improve this, but maybe @bruceg has any thoughts?

jszwedko commented on lib/vrl/stdlib/src/parse_json.rs:
```diff
@@ -7,6 +10,63 @@ fn parse_json(value: Value) -> Resolved {
+    let res = value.try_integer()?;
+    if !(1..=128).contains(&res) {
+        Err(ExpressionError::from(format!(
+            "max_depth value should be greater than 0 and less than 128, got {}",
```
Could we mention this range restriction in the docs for this config option? I'm also wondering why we need to cap it at all?

nabokihms replied:
The lower cap is `1` because it is pointless to use anything lover, `data = parse_json!(.message, max_depth: 0)` equals to `data = .message`.

The upper cap is 128 because serde_json [has the same](https://github.com/serde-rs/json/blob/4d57ebeea8d791b8a51c229552d2d480415d00e6/json/src/de.rs#L111) recursion limit by default.

I will add the info to the doc.

JeanMertz commented on lib/vrl/stdlib/benches/benches.rs:
```diff
@@ -1418,6 +1418,11 @@ bench_function! {
         want: Ok(value!({key: "value"})),
     }
+
+    map_max_depth {
+        args: func_args![value: r#"{"key": "value"}"#, max_depth: 128],
+        want: Ok(value!({key: "value"})),
+    }
```
Could you add another benchmark, that actually iterates over, say 5 nested objects/arrays?  Because right now, it does look like the performance is 2,5x worse, without it actually doing any recursion, so I'm curious what happens to those numbers if we _do_ let it recurse.

bruceg commented on lib/vrl/stdlib/src/parse_json.rs:
```diff
@@ -162,5 +263,41 @@ mod tests {
                 .add_object(Collection::from_unknown(inner_kind())),
         }
+
+        max_depth {
+            args: func_args![ value: r#"{"top_layer": {"layer_one": "finish", "layer_two": 2}}"#, max_depth: 1],
+            want: Ok(value!({ top_layer: r#"{"layer_one": "finish", "layer_two": 2}"# })),
+            tdef: type_def(),
+        }
+
+        max_depth_array {
+            args: func_args![ value: r#"[{"top_layer": {"next_layer": ["finish"]}}]"#, max_depth: 2],
+            want: Ok(value!([{ top_layer: r#"{"next_layer": ["finish"]}"# }])),
+            tdef: type_def(),
+        }
+
+        max_depth_exhausted {
+            args: func_args![ value: r#"{"top_layer": {"layer_one": "finish", "layer_two": 2}}"#, max_depth: 10],
+            want: Ok(value!({ top_layer: {layer_one: "finish", layer_two: 2} })),
+            tdef: type_def(),
+        }
```
None of these tests appear to actually test what happens when the `max_depth` is exhausted.

nabokihms replied:
Yeah, I meant that the max_depth value exceeds the number of the JSON document. I will fix the naming.

bruceg commented on lib/vrl/stdlib/src/parse_json.rs:
```diff
@@ -7,6 +10,63 @@ fn parse_json(value: Value) -> Resolved {
+) -> std::result::Result<JsonValue, Error> {
+    if value.get().starts_with('{') {
+        if max_depth == 0 {
+            serde_json::value::to_value(value.to_string())
```
Does this actually work? This leaves me all kinds of confused what is going on here, so maybe some code comments would help explain the logic.

nabokihms replied:
Sure, no problem. I will try to clarify the logic.

nabokihms commented:
The remaining status:

1. I added more comments to the code (some may be too obvious). Let me know if it is so.
2. The doc is enriched with more information about max_depth borders (according to @jszwedko request). I also added more comments to the code about the validation.
3. Benchmarks, oh benchmarks... I managed to improve performance a little by migrating from `Box<RawValue>` to `<&RawValue>`.

The current results on my PC are:
```
vrl_stdlib/functions/parse_json/map                                                                            
                        time:   [446.14 ns 455.42 ns 464.77 ns]
                        thrpt:  [2.1516 Melem/s 2.1958 Melem/s 2.2414 Melem/s]

vrl_stdlib/functions/parse_json/map_max_depth                                                                             
                        time:   [937.12 ns 952.49 ns 968.91 ns]
                        thrpt:  [1.0321 Melem/s 1.0499 Melem/s 1.0671 Melem/s]

vrl_stdlib/functions/parse_json/nested                                                                             
                        time:   [1.9106 us 1.9494 us 1.9904 us]
                        thrpt:  [502.42 Kelem/s 512.98 Kelem/s 523.40 Kelem/s]

vrl_stdlib/functions/parse_json/nested_max_depth_1                                                                             
                        time:   [1.1976 us 1.2178 us 1.2386 us]
                        thrpt:  [807.38 Kelem/s 821.18 Kelem/s 834.99 Kelem/s]

vrl_stdlib/functions/parse_json/nested_max_depth_10                                                                             
                        time:   [5.1829 us 5.2465 us 5.3187 us]
                        thrpt:  [188.02 Kelem/s 190.60 Kelem/s 192.94 Kelem/s]
```

bruceg commented on lib/vrl/stdlib/src/parse_json.rs:
```diff
@@ -162,5 +281,41 @@ mod tests {
                 .add_object(Collection::from_unknown(inner_kind())),
         }
+
+        max_depth {
+            args: func_args![ value: r#"{"top_layer": {"layer_one": "finish", "layer_two": 2}}"#, max_depth: 1],
+            want: Ok(value!({ top_layer: r#"{"layer_one": "finish", "layer_two": 2}"# })),
+            tdef: type_def(),
+        }
+
+        max_depth_array {
+            args: func_args![ value: r#"[{"top_layer": {"next_layer": ["finish"]}}]"#, max_depth: 2],
+            want: Ok(value!([{ top_layer: r#"{"next_layer": ["finish"]}"# }])),
+            tdef: type_def(),
+        }
+
+        max_depth_exceeds_layers {
+            args: func_args![ value: r#"{"top_layer": {"layer_one": "finish", "layer_two": 2}}"#, max_depth: 10],
+            want: Ok(value!({ top_layer: {layer_one: "finish", layer_two: 2} })),
+            tdef: type_def(),
+        }
```
I'm looking for a test that shows that `max_depth` affects the output. If I'm reading this right, the depth limit is not changing the output in any of these tests.

bruceg replied:
Never mind, I misread the quoted bits in the output lines. I was expecting to see the inner values dropped and not left as strings.

JeanMertz commented on lib/vrl/stdlib/src/parse_json.rs:
```diff
@@ -85,13 +175,24 @@ impl Function for ParseJson {
+        let max_depth = arguments.optional("max_depth");
 
-        Ok(Box::new(ParseJsonFn { value }))
+        if let Some(max_depth) = max_depth {
+            Ok(Box::new(ParseJsonMaxDepthFn { value, max_depth }))
+        } else {
+            Ok(Box::new(ParseJsonFn { value }))
+        }
```
nit

```suggestion
        match max_depth {
            Some(max_depth) => Ok(Box::new(ParseJsonMaxDepthFn { value, max_depth })),
            None => Ok(Box::new(ParseJsonFn { value })),
        }
```

JeanMertz reviewed (approved):
Nice! Thank you.

bruceg commented on lib/vrl/stdlib/src/parse_json.rs:
```diff
@@ -162,5 +280,41 @@ mod tests {
+        }
+
+        max_depth_array {
+            args: func_args![ value: r#"[{"top_layer": {"next_layer": ["finish"]}}]"#, max_depth: 2],
+            want: Ok(value!([{ top_layer: r#"{"next_layer": ["finish"]}"# }])),
```
I am also a little confused about the layer counting here. Should this output two layers:
```suggestion
            args: func_args![ value: r#"[{"top_layer": {"next_layer": ["finish"]}}]"#, max_depth: 2],
            want: Ok(value!([{ top_layer: {"next_layer": r#"["finish"]"# }}])),
```

nabokihms replied:
In this example, the first layer is the array, and the second one is the object. I can swap them to make it more obvious or add a comment.

bruceg replied:
Ah, okay, that's what I was missing.

bruceg reviewed (commented):
I have a couple questions about the level counting in the test cases, but I'm good with the rest of it too.

nabokihms commented on lib/vrl/stdlib/benches/benches.rs:
```diff
@@ -1418,6 +1418,26 @@ bench_function! {
+        want: Ok(value!({"1":{"2":{"3":{"4":{"5":{"6":"end"}}}}}})),
+    }
+
+    nested_max_depth_1 {
+        args: func_args![value: r#"{"1":{"2":{"3":{"4":{"5":{"6":"end"}}}}}}"#, max_depth: 1],
+        want: Ok(value!({"1":{"2":"{\"3\":{\"4\":{\"5\":{\"6\":\"end\"}}}}"}})),
+    }
```
Yeah, you are right. My bad.

bruceg reviewed (approved):
Thanks for this addition, @nabokihms
